src/common/hash.h

```diff
@@ -68,4 +70,13 @@ struct HashableStruct {
     }
 };
 
+struct PairHash {
+    template <class T1, class T2>
+    std::size_t operator()(const std::pair<T1, T2>& pair) const {
```

Comment:
```suggestion
    std::size_t operator()(const std::pair<T1, T2>& pair) const noexcept {
```

Hashes shouldn't throw exceptions (and standard facilities may check whether the hash will throw or not when determining which code path to use).